src/resolve/resolved-dns-browse-services.c

```diff
@@ -0,0 +1,710 @@
+                                                JSON_BUILD_PAIR("family", JSON_BUILD_INTEGER(owner_family)),
+                                                JSON_BUILD_PAIR("name", JSON_BUILD_STRING(name)),
+                                                JSON_BUILD_PAIR("type", JSON_BUILD_STRING(type)),
+                                                JSON_BUILD_PAIR("domain", JSON_BUILD_STRING(domain)),
+                                                JSON_BUILD_PAIR("interface", JSON_BUILD_INTEGER(sb->ifindex))));
+                if (r < 0)
+                        goto finish;
+
+                r = json_variant_append_array(&array, entry);
+                if (r < 0)
+                        goto finish;
+        }
+
+        if (!json_variant_is_blank_array(array)) {
+                _cleanup_(json_variant_unrefp) JsonVariant *vm = NULL;
+
+                r = json_build(&vm,
+                                JSON_BUILD_OBJECT(
+                                                JSON_BUILD_PAIR("browser_service_data", JSON_BUILD_VARIANT(array))));
+                if (r < 0)
+                        goto finish;
+
+                r = varlink_notify(sb->link, vm);
+                if (r < 0)
+                        goto finish;
+        }
+
+        return 0;
+
+finish:
+        log_error_errno(r, "Failed to process received services: %m");
+        return varlink_error_errno(sb->link, r);
+}
+
+int mdns_browser_lookup_cache(DnsServiceBrowser *sb, int owner_family) {
+        _cleanup_(dns_answer_unrefp) DnsAnswer *lookup_ret_answer = NULL;
+        DnsScope *scope;
+        int r;
+
+        assert(sb);
+        assert(sb->m);
+
+        scope = manager_find_scope_from_protocol(sb->m, sb->ifindex, DNS_PROTOCOL_MDNS, owner_family);
+        if (!scope)
+                return 0;
+
+        dns_cache_prune(&scope->cache);
+
+        r = dns_cache_lookup(
+                        &scope->cache,
+                        sb->key,
+                        sb->flags,
+                        NULL,
+                        &lookup_ret_answer,
+                        NULL,
+                        NULL,
+                        NULL);
+        if (r < 0)
+                return r;
+
+        return mdns_manage_services_answer(sb, lookup_ret_answer, owner_family);
+}
+
+static int goodbye_callback(sd_event_source *s, uint64_t usec, void *userdata) {
+        _cleanup_(mdns_goodbye_params_freep) mDnsGoodbyeParams *cb_params = userdata;
+        int r;
+
+        if (!cb_params->sb)
+                return 0;
+
+        r = mdns_browser_lookup_cache(cb_params->sb, cb_params->owner_family);
+        if (r < 0)
+                return log_error_errno(r, "Failed to lookup cache: %m");
+
+        return 0;
+}
+
+int mdns_notify_browsers_unsolicited_updates(Manager *m, DnsAnswer *answer, int owner_family, bool has_goodbye) {
+        _cleanup_(dns_service_browser_unrefp) DnsServiceBrowser *sb = NULL;
+        int r;
+
+        assert(answer);
```

Comment:
I haven't looked at the PR closely but it looks like resolved can still be crashed by sending unsolicited responses with no answers:
```sh
Received mdns UDP packet of size 12, ifindex=2, ttl=1, fragsize=0, sender=192.168.93.177, destination=224.0.0.251
Got mDNS reply packet
Assertion 'answer' failed at src/resolve/resolved-dns-browse-services.c:431, function mdns_notify_browsers_unsolicited_updates(). Aborting.
systemd-resolved.service: Main process exited, code=dumped, status=6/ABRT
systemd-resolved.service: Failed with result 'core-dump'.
systemd-resolved.service: Scheduled restart job, restart counter is at 7.
systemd-resolved.service: Start request repeated too quickly.
systemd-resolved.service: Failed with result 'core-dump'.
Failed to start systemd-resolved.service.
```
```sh
#0  0x00007f48d80b0884 in __pthread_kill_implementation () from target:/lib64/libc.so.6
#1  0x00007f48d805fafe in raise () from target:/lib64/libc.so.6
#2  0x00007f48d804887f in abort () from target:/lib64/libc.so.6
#3  0x00007f48d9a38c11 in log_assert_failed (text=0x79ee80 <str> "answer", file=0x79fbd7 "src/resolve/resolved-dns-browse-services.c", line=431, 
    func=0x79eea0 <__func__.mdns_notify_browsers_unsolicited_updates> "mdns_notify_browsers_unsolicited_updates") at ../src/basic/log.c:967
#4  0x00000000007337d3 in mdns_notify_browsers_unsolicited_updates (m=0x619000000580, answer=0x0, owner_family=2, has_goodbye=false)
    at ../src/resolve/resolved-dns-browse-services.c:431
#5  0x000000000062e00c in on_mdns_packet (s=0x60f000002980, fd=15, revents=1, userdata=0x619000000580) at ../src/resolve/resolved-mdns.c:457
#6  0x00007f48da03320d in source_dispatch (s=0x60f000002980) at ../src/libsystemd/sd-event/sd-event.c:4187
#7  0x00007f48da031235 in sd_event_dispatch (e=0x617000000080) at ../src/libsystemd/sd-event/sd-event.c:4808
#8  0x00007f48da036d2e in sd_event_run (e=0x617000000080, timeout=18446744073709551615) at ../src/libsystemd/sd-event/sd-event.c:4869
#9  0x00007f48da037cba in sd_event_loop (e=0x617000000080) at ../src/libsystemd/sd-event/sd-event.c:4891
#10 0x0000000000669759 in run (argc=1, argv=0x7ffe99235028) at ../src/resolve/resolved.c:92
#11 0x000000000066779e in main (argc=1, argv=0x7ffe99235028) at ../src/resolve/resolved.c:99
```

Reply:
The packet was
```sh
Multicast Domain Name System (response)
    Transaction ID: 0x0000
    Flags: 0x8100 Standard query response, No error
        1... .... .... .... = Response: Message is a response
        .000 0... .... .... = Opcode: Standard query (0)
        .... .0.. .... .... = Authoritative: Server is not an authority for domain
        .... ..0. .... .... = Truncated: Message is not truncated
        .... ...1 .... .... = Recursion desired: Do query recursively
        .... .... 0... .... = Recursion available: Server can't do recursive queries
        .... .... .0.. .... = Z: reserved (0)
        .... .... ..0. .... = Answer authenticated: Answer/authority portion was not authenticated by the server
        .... .... ...0 .... = Non-authenticated data: Unacceptable
        .... .... .... 0000 = Reply code: No error (0)
    Questions: 0
    Answer RRs: 0
    Authority RRs: 0
    Additional RRs: 0
    [Unsolicited: True]
```
```python
>>> raw(p2[DNS])
b'\x00\x00\x81\x00\x00\x00\x00\x00\x00\x00\x00\x00'
```

Reply:
Thanks for catching this.
I don't know when the assert crept back in. Replaced it with a null check.

Reply:
Looks like it no longer crashes there.